src/named_pipe.rs

```diff
@@ -0,0 +1,208 @@
+        {
+            Ok((c_in.as_raw_fd(), c_out.as_raw_fd()))
+        } else {
+            Err(WorkerErrorKind::OperationTimedOut.into())
+        }
+    }
+
+    pub fn write(&mut self, v: &[u8]) -> Result<(), WorkerError> {
+        let (c_in_fd, _) = self.get_fds()?;
+
+        let deadline = Instant::now() + Duration::from_millis(PIPE_IO_TIMEOUT_MS);
+
+        assert_eq!(v.last(), Some(&b'\n'));
+
+        let mut write_idx = 0;
+        while write_idx < v.len() && Instant::now() < deadline {
+            // Wait until ready
+            let poll_flags = PollFlags::POLLOUT;
+            poll(
+                &mut [PollFd::new(c_in_fd, poll_flags)],
+                (deadline - Instant::now()).as_millis() as c_int,
+            )?;
+
+            // Then write the bytes
+            let written_bytes = write(c_in_fd, &v[write_idx..])?;
+            write_idx += written_bytes;
+        }
+
+        // If we didn't write everything, we timed out
+        if write_idx < v.len() {
+            return Err(WorkerErrorKind::OperationTimedOut.into());
+        }
+
+        Ok(())
+    }
+
+    pub fn component_input_file(&self) -> &PathBuf {
+        &self.component_input_fifo_path
+    }
+
+    pub fn read(&mut self) -> Result<Vec<u8>, WorkerError> {
+        let (_, c_out_fd) = self.get_fds()?;
+
+        let deadline = Instant::now() + Duration::from_millis(PIPE_IO_TIMEOUT_MS);
+
+        // Then read the bytes
+        let mut read_buf = vec![0; BUF_SIZE];
+        let mut result = Vec::with_capacity(BUF_SIZE);
+        loop {
+            // Wait for data to be available
+            debug!("Polling {:?}", self.component_output_fifo_path);
+            let poll_flags = PollFlags::POLLIN;
+            poll(
+                &mut [PollFd::new(c_out_fd, poll_flags)],
+                (deadline - Instant::now()).as_millis() as c_int,
+            )?;
+
+            // If we've timed out, then just return an error
+            if Instant::now() > deadline {
+                return Err(WorkerErrorKind::OperationTimedOut.into());
+            }
+
+            // Otherwise read n bytes
+            let n = match read(c_out_fd, &mut read_buf) {
+                Ok(n) => n,
+                Err(e) => {
+                    if e.as_errno() == Some(Errno::EAGAIN) {
+                        debug!("Trying again");
+                        sleep(Duration::from_millis(PIPE_POLL_INTERVAL_MS));
+                        continue;
+                    } else {
+                        return Err(e.into());
+                    }
+                }
+            };
+
+            // Reading 0 bytes indicates unix doesn't think there is more to read
+            if n == 0 {
+                return Err(WorkerErrorKind::SubprocessDisconnected.into());
+            }
+
+            for &v in &read_buf[0..n] {
+                result.push(v);
+                if v == b'\n' {
+                    return Ok(result);
+                }
```

Comment:
We should verify this is the last byte

Reply:
(I've decided not to do this, since we actually will catch this on the next request)